libraries/cms/Provider/ApplicationProvider.php

```diff
@@ -0,0 +1,97 @@
+<?php
+/**
+ * @package     Joomla.Libraries
+ * @subpackage  Service
+ *
+ * @copyright   Copyright (C) 2005 - 2015 Open Source Matters, Inc. All rights reserved.
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+namespace Joomla\Cms\Provider;
+
+defined('JPATH_PLATFORM') or die;
+
+use Joomla\DI\Container;
+use Joomla\DI\ServiceProviderInterface;
+
+/**
+ * The CMS application service provider.
+ *
+ * @since  4.0
+ */
+class ApplicationProvider implements ServiceProviderInterface
+{
+	public function register(Container $container)
+	{
+		$container->set('JApplicationSite', array($this, 'getApplicationSite'));
+		$container->set('JApplicationAdministrator', array($this, 'getApplicationAdministrator'));
+		$container->set('InstallationApplicationWeb', array($this, 'getApplicationInstallation'));
```

Comment:
Typically we've kept the install app isolated from the rest of the code structure.  Not 100% sure on whether we should add this here yet.